### DISCUSSION TITLE
Use a proxy as context for granular change detection

### DISCUSSION THREAD

**Discussion by geekox86** [statelyai/xstate#2452] [Q&A] [answered]
Hello everyone. How can I use a JS proxy as my machine context?

I need it for granular change detection of the context. While `onChange` is there, I want to avoid the overhead of diffing.

If currently not possible, may I request it to be considered as a new feature in v5?

**Comment by Andarist** [comment, marked as answer]
I don't think this is currently supported. It would be interesting to explore this area - it potentially would be very useful for integrations with frameworks like SolidJS.

**Reply by geekox86** [reply]
It can also enable declarative computed/derived fields.

**Comment by davidkpiano** [comment]
> I need it for granular change detection of the context. 

Can you show some example (pseudo-)code of this?

**Reply by geekox86** [reply]
> Can you show some example (pseudo-)code of this?
```
const machine = createMachine(
	{
		context: new Proxy(
			{
				array: [1, 2, 3]
			},
			{
				get: (target, prop, receiver) => {
					console.log("Dependency detected for ", prop);

					// register dependencies here

					return Reflect.get(target, prop, receiver);
				},
				set: (target, prop, value) => {
					console.log("Dependency update detected for ", prop);

					// notify dependants here

					return Reflect.set(target, prop, value);
				}
			}
		),
		initial: "initializing",
		states: {
			initializing: {}
		}
	}
);
```

**Reply by geekox86** [reply]
Please, also take a look on my other related discussion #2453 regarding context initialization and privacy.
